crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -460,12 +460,39 @@ def check_schema_migrations
         false
       end
 
+      # Check the version of installed SLES to verify admin upgrade went well.
+      # We have to do the check here at the start of 'services' step and not during admin upgrade,
+      # because admin upgrade step is not repeatable.
+      def check_product_version
+        zypper_stream = Hash.from_xml(
+          `sudo /usr/bin/zypper-retry --xmlout products -i`
+        )["stream"]
+        products = zypper_stream["product_list"]["product"]
+        return true if products.detect do |product|
```

Comment:
1. it will die if it doesn't find anything as `detect` will return `nil` and you can't call `any?` on `nil`
2. you don't need `detect` at all. you just want to check presence of "any" item matching your predicate so something like:
```
return true if products.any? { |p| p["name"] == "SLES" && p["version"] == "12.4" }
```
should do the trick.

Reply:
True.